### PR TITLE
Feature/spack stack extension

### PR DESCRIPTION
**Pull request by kgerheiser**
Depends on https://github.com/NOAA-EMC/spack/pull/63

`./create environment/container` becomes `spack stack create env/container`. And instead of `--config` it's a positional argument as the first arg to `create container`.

The CI is running now and I was able to generate a valid container. The runs havemade it through the environment setup, so all seems well.

### CODE REVIEW

File: configs/apps/ufs-weather-model/spack.yaml [hunk 1]
```diff
@@ -1,9 +1,8 @@
-# Template spack.yaml for apps
 spack:
   concretization: separately
   view: false
 
-  @CONFIG_INCLUDES@
+  include: []
```
Comment (climbfuji) [inline configs/apps/ufs-weather-model/spack.yaml:5]: Why does only the ufs-weather-model have `include: []`, but none of the other apps?

File: configs/containers/docker-ubuntu-gcc-openmpi.yaml [hunk 2]
```diff
@@ -114,5 +97,5 @@ spack:
 
     # Labels for the image
     labels:
-      app: "@SPEC@"
+      app: ""
```
Comment (climbfuji) [inline configs/containers/docker-ubuntu-gcc-openmpi.yaml:100]: Not sure if this label is needed at all (that came from the spack template for containers), but if it is, is there a way to set it to the spec?

**Review by climbfuji — COMMENTED** [review]
This is grea, thanks also for updating the documentation! Just a couple of questions for my understanding.

File: configs/containers/docker-ubuntu-gcc-openmpi.yaml (hunk 2 above)
Comment (kgerheiser) [reply configs/containers/docker-ubuntu-gcc-openmpi.yaml:100]: Yes, I meant to add this. Just a key insert.

File: configs/apps/ufs-weather-model/spack.yaml (hunk 1 above)
Comment (kgerheiser) [reply configs/apps/ufs-weather-model/spack.yaml:5]: No good reason. Something I was playing with how the yaml keys were being inserted. If they key exists it affects the ordering of how new things are written, so for visual style I was trying things out.

File: configs/containers/docker-ubuntu-gcc-openmpi.yaml (hunk 2 above)
Comment (kgerheiser) [reply configs/containers/docker-ubuntu-gcc-openmpi.yaml:100]: Fixed

File: configs/apps/ufs-weather-model/spack.yaml (hunk 1 above)
Comment (kgerheiser) [reply configs/apps/ufs-weather-model/spack.yaml:5]: I added the include everywhere just so the includes placed near the top of the file.

**Comment by climbfuji** [comment]
Great, I am going to test this on my Mac - interested to see the order of includes (and how spack overwrites settings).

**Comment by kgerheiser** [comment]
That's a really good point that I should mention. The precedence from lowest to highest is base packages (i.e. `common/packages.yaml` unless overridden or excluded), then site, then what's in the `spack.yaml`. I think this is a logical way to handle it. Then you can have your base packages, and a special environment that explicitly asks for something different. by adding it to its `spack.yaml`.

**Review by climbfuji — COMMENTED** [review]
A few questions from testing this out:
1. `spack stack env help` isn't up to date I think
2. If no site is specified it defaults to `default`, not `none`
3. Can we hide the unused/unnecessarily complex and confusing arguments for `spack stack env create`? (possibly the same for containers)
4. I noticed the site include dir is now called `site.default` (or `site.something-else`), no longer just `site`. While this is not a big problem, was this intentional? I liked the previous `site` for its simplicity.

**Comment by climbfuji** [comment]
> That's a really good point that I should mention. The precedence from lowest to highest is base packages (i.e. `common/packages.yaml` unless overridden or excluded), then site, then what's in the `spack.yaml`. I think this is a logical way to handle it. Then you can have your base packages, and a special environment that explicitly asks for something different. by adding it to its `spack.yaml`.

Excellent. I checked this on my Mac, and packages such as `fms` that have an override of the OpenMP settings (`~openmp`) in the site config are configured correctly.

**Comment by kgerheiser** [comment]
> spack stack env help isn't up to date I think

You're right

> If no site is specified it defaults to default, not none

What's the correct behavior here? It can either default to default, or default to none. I went back and forth.

> Can we hide the unused/unnecessarily complex and confusing arguments for spack stack env create? (possibly the same for containers)

Yes, I added some of these for some weird corner cases that are probably not necessary.

> I noticed the site include dir is now called site.default (or site.something-else), no longer just site. While this is not a big problem, was this intentional? I liked the previous site for its simplicity.

It was intentional, to make it more clear. I can just put it back to `site`.

**Comment by climbfuji** [comment]
> > spack stack env help isn't up to date I think
> 
> You're right
> 
> > If no site is specified it defaults to default, not none
> 
> What's the correct behavior here? It can either default to default, or default to none. I went back and forth.

default is the correct behavior, but I think the help string or documentation isn't correct.

> 
> > Can we hide the unused/unnecessarily complex and confusing arguments for spack stack env create? (possibly the same for containers)
> 
> Yes, I added some of these for some weird corner cases that are probably not necessary.

Maybe we comment them out until we need them?

> 
> > I noticed the site include dir is now called site.default (or site.something-else), no longer just site. While this is not a big problem, was this intentional? I liked the previous site for its simplicity.
> 
> It was intentional, to make it more clear. I can just put it back to `site`.

No, that's fine, I just wanted to know.

**Comment by kgerheiser** [comment]
Updated the help of container and env with a dynamic list of available options.

Removed some of the CLI args for simplicity, and combined the common argarse options.

Re-named site.<name>, to just site.

File: README.md [hunk 3]
```diff
@@ -81,13 +81,13 @@ spack module lmod refresh
 ### Create container
 ```
 # See a list of preconfigured containers
-./create.py container -h
+spack stack create container -h
 
 # Create container spack definition (spack.yaml) in directory envs/<spec>.<config>
-./create.py container --config docker-ubuntu-gcc-openmpi --spec esmf
+spack stack create container docker-ubuntu-gcc-openmpi --app ufs-weather-model
```
Comment (climbfuji) [inline README.md:87]:
```suggestion
spack stack create container docker-ubuntu-gcc-openmpi.yaml --app ufs-weather-model
```

Comment (climbfuji) [reply README.md:87]: I just checked and it now wants the extension as part of the container name. This was probably not intentional.

**Comment by climbfuji** [comment]
@kgerheiser The container creation spits out a nice message at the end:
```
==> Created container /Users/heinzell/scratch/spack-stack-feature-spack-stack-extension-final/envs/docker-ubuntu-gcc-openmpi.yaml.jedi-ewok
```
Would be nice if we could do the same for environments.

File: README.md (hunk 3 above)
Comment (kgerheiser) [reply README.md:87]: Do you have an updated branch? This was a problem that I fixed. I can create a container with no `.yaml` suffix.

**Comment by kgerheiser** [comment]
Added a message to the env create.

File: README.md (hunk 3 above)
Comment (climbfuji) [reply README.md:87]: I saw you just pushed that (https://github.com/NOAA-EMC/spack/pull/63/commits/c8e99bc1a533db38b4eac7da56f1c2fbb385f816)

**Review by climbfuji — CHANGES_REQUESTED** [review]
I am now looking at the container builds in a little more detail, and I noticed a few differences (positive or negative):
1. Negative/doesn't matter? Existing container build uses concretization `together`, now it uses `separately` and spack spits out a warning that `together` is recommended.
2. Negative. The `packages:all` section now has the problem with the double colons. I worked around this by replacing double colons with single colons when replacing @PACKAGE_CONFIG@ with the actual contents of the common `packages.yaml`. Now it is:
```
 # Basic package config from configs/common/packages.yaml
  # Additional package config for container
  packages:
    all:
      providers:
        'blas:': [openblas]
        'lapack:': [openblas]
        'fftw-api:': [fftw]
        'yacc:': [bison]
        'gl:': [opengl]
        'glu:': [openglu]
        'jpeg:': [libjpeg-turbo]
```
3. Positive: all the comments in the common `packages.yaml` are stripped out in the container `spack.yaml`, and long lists of `variants` are formatted for better readability.
4. Negative/doesn't matter? The `app` string is still empty:
```
    # Labels for the image
    labels:
      app: ''
      mpi: openmpi
```

**Comment by kgerheiser** [comment]
> Negative/doesn't matter? Existing container build uses concretization together, now it uses separately and spack spits out a warning that together is recommended.

Yes, I'll fix that.

> Negative. The packages:all section now has the problem with the double colons. I worked around this by replacing double colons with single colons when replacing @PACKAGE_CONFIG@ with the actual contents of the common packages.yaml.

Yes, there is a problem when handling double colons. I'll do what you did and strip the double colons. Would be nice if Spack could handle this properly. It works everywhere else except spack.yaml (because it's not a real config file).

> Negative/doesn't matter? The app string is still empty:

I have code that handles this and it's working for me. I get:

```
 labels:
      app: ufs-weather-model
      mpi: openmpi
```

**Comment by climbfuji** [comment]
> > Negative/doesn't matter? Existing container build uses concretization together, now it uses separately and spack spits out a warning that together is recommended.
> 
> Yes, I'll fix that.
> 
> > Negative. The packages:all section now has the problem with the double colons. I worked around this by replacing double colons with single colons when replacing @PACKAGE_CONFIG@ with the actual contents of the common packages.yaml.
> 
> Yes, there is a problem when handling double colons. I'll do what you did and strip the double colons. Would be nice if Spack could handle this properly. It works everywhere else except spack.yaml (because it's not a real config file).
> 
> > Negative/doesn't matter? The app string is still empty:
> 
> I have code that handles this and it's working for me. I get:
> 
> ```
>  labels:
>       app: ufs-weather-model
>       mpi: openmpi
> ```

Thanks, I'll try the labels again once everything is updated. Maybe the submodule pointer for `spack` wasn't updated to your latest code when I checked out the code 2hrs ago.

**Comment by kgerheiser** [comment]
Double colons, labels, and concretization (merge precedence) has been fixed, and some unit tests added that loop over each app, and site checking that it works.

`spack unit-test -s --extension=stack`

File: .spackstack [hunk 4]
```diff
@@ -0,0 +1 @@
+Let's spack stack command ffind top-level spack-stack directory for config location
```
Comment (climbfuji) [inline .spackstack]:
```suggestion
Lets spack stack command find top-level spack-stack directory for config location
```

File: configs/apps/empty/spack.yaml [hunk 5]
```diff
@@ -2,6 +2,6 @@ spack:
   concretization: separately
   view: false
   
-  @CONFIG_INCLUDES@
+
```
Comment (climbfuji) [inline configs/apps/empty/spack.yaml]:
```suggestion
  include: []
```

**Review by climbfuji — COMMENTED** [review]
Did the CI tests pass after the spack PR was merged?

**Comment by kgerheiser** [comment]
Let's see - Ubuntu-gcc timed out at 6h on ewok (all other builds worked), and Intel failed because of some apt error.

I'll start them again because there were a decent number of changes since, and I didn't run macOS at all.

**Review by climbfuji — APPROVED** [review]
I'll approve this now so that you can merge when the test are passing and you are happy with it.

**Comment by kgerheiser** [comment]
The GCC build works Along with the unit tests, the builds are all chugging along and no actual build settings were changed. Once the environment is created it's all Spack. 

Next up I'm going to move the meta module Python code into a custom command, which should be comparatively simple.

**Comment by climbfuji** [comment]
> The GCC build works Along with the unit tests, the builds are all chugging along and no actual build settings were changed. Once the environment is created it's all Spack.
> 
> Next up I'm going to move the meta module Python code into a custom command, which should be comparatively simple.

Thanks! Yes, the meta module update and my doc update should be the last changes needed before the first release imo.

**Comment by climbfuji** [comment]
I added the ability to detect qt5 to spack yesterday, we should be able to use that to speed up the ewok builds. I think qt5 gets installed already via brew / apt, but isn’t used yet (spack builds it if I am not mistaken).

**Pull request merged**